### ISSUE TITLE
shared_examples/shared_context user defined metadata doesn't override existsing one

### ISSUE THREAD

**Issue by ojab** [rspec/rspec-core#2306]
Consider the following spec:
```ruby
RSpec.shared_examples 'shared examples', foo: 'baz' do
  it { 'Works!' }
end

RSpec.feature 'feature', foo: 'bar' do
  before(:each, foo: 'bar') { raise }
  it_behaves_like 'shared examples'

  it 'works', foo: 'baz' do
  end
  context 'inner', foo: 'baz' do
    it 'works too' do
    end
  end
end

RSpec.feature 'another feature' do
  before(:each, foo: 'baz') { raise }
  it_behaves_like 'shared examples'
end
```
'shared examples' in 'feature' should work fine as other examples do, since inner user defined metadata overrides feature's one.
As can be seen by failure of 'another feature' -- metadata is indeed set by 'shared examples'.

I assume that the current (rspec-core-3.5.1) behavior is a bug that should be fixed.

**Comment by myronmarston** [comment]
What config value are you using for `config.shared_context_metadata_behavior`?  It's new in 3.5 and determines how metadata for shared contexts behave.  What behavior is expected depends on how you've set that.

**Comment by ojab** [comment]
Tried with both `:trigger_inclusion` and `:apply_to_host_groups`, included examples in 'feature' always fails.
Behavior of 'another feature' changes, it works with `:trigger_inclusion`, but fails with `:apply_to_host_groups`.

**Comment by myronmarston** [comment]
Just to make sure we're clear about expectations, can you paste the verbatim output you are getting and an example of the output you expect?

**Comment by ojab** [comment]
I assume that `:trigger_inclusion` should not be changed anyway because of deprecation.
And after reading 3.5.0 release notes (about `:apply_to_host_groups`) I actually not so sure about desired behavior :/.
Right now (w/ `:apply_to_host_groups`) it is
```
feature
  works
  behaves like shared examples
    example at ./spec/xxx_spec.rb:2 (FAILED - 1)
  inner
    works too

another feature
  behaves like shared examples
    example at ./spec/xxx_spec.rb:2 (FAILED - 2)
```
'another feature' failure is expected, but personally I expect it to be the only failure, so output should be:
```
feature
  works
  behaves like shared examples
    example at ./spec/xxx_spec.rb:2
  inner
    works too

another feature
  behaves like shared examples
    example at ./spec/xxx_spec.rb:2 (FAILED - 2)
```

AFAIU `:apply_to_host_groups` leads to `foo: 'baz'` metadata from 'shared examples'  to be added to 'feature', overriding `foo: 'bar'`.
IMO in case of `it_behaves_like` (and other methods that add the nested context) metadata should be applied to this nested context, not parent group. So right now nesting w/ regards to metadata looks like
```
feature
  shared_examples
    nested_context
```
while it would be more flexible (and more reasonable in my view) to be like
```
feature
  nested_context
     shared_examples
```
i. e. if shared_examples metadata would be propagated only to the nested_context, not to the feature.

`include_examples` and other methods that do not add a nested context should work without changes, because
```
feature
  shared_examples
```
leads to metadata propagation to the upper/feature level.

Hope that my description is clear enough.

**Comment by myronmarston** [comment]
Thanks, that's very helpful.

When a shared example group is included, we decided that it should not stomp metadata previously set on the group -- the idea being that on a specific group you know what metadata you want, whereas on a shared group, you don't know about the needs of all inclusions, so we thought it made sense to give the host group precedence over the same metadata key on an included shared group.

Now, the only case I had in mind when writing that feature was `include_context` -- where the contents of the shared group are directly included (and the metadata is directly applied to the host group).  For your case you are using `it_behaves_like` which creates a nested group, and there's a much stronger argument to be made that since it creates a nested group, and nested group metadata takes precedence over metadata inherited from an outer group, the shared metadata should take metadata for that case.

I'll see if it's feasible to implement that.

**Comment by myronmarston** [comment]
I have a fix ready in #2307 if you want to give it a try.
